core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java

```diff
@@ -71,12 +71,18 @@ public void onCommit(List<? extends IndexCommit> commits) throws IOException {
     }
 
     private void setLastCommittedTranslogGeneration(List<? extends IndexCommit> commits) throws IOException {
-        // when opening an existing lucene index, we currently always open the last commit.
-        // we therefore use the translog gen as the one that will be required for recovery
-        final IndexCommit indexCommit = commits.get(commits.size() - 1);
-        assert indexCommit.isDeleted() == false : "last commit is deleted";
-        long minGen = Long.parseLong(indexCommit.getUserData().get(Translog.TRANSLOG_GENERATION_KEY));
-        translogDeletionPolicy.setMinTranslogGenerationForRecovery(minGen);
+        // We need to keep translog since the smallest translog generation of un-deleted commits.
+        // However, there are commits that are not deleted just because they are being snapshotted (rather than being kept by the policy).
+        // TODO: We need to distinguish those commits and skip them in calculating the minimum required translog generation.
+        long minRequiredGen = Long.MAX_VALUE;
+        for (IndexCommit indexCommit : commits) {
```

Comment:
you can do some streaming java8 magic here.

Reply:
Yep, but `indexCommit.getUserData()` throws IOException.